File: src/lib/format.ts
const SYMBOLS: Record<string, string> = {
  EUR: "\u20ac", USD: "$", GBP: "\u00a3", CHF: "CHF ", SEK: "kr ", NOK: "kr ",
  DKK: "kr ", PLN: "z\u0142 ", CZK: "K\u010d ", HUF: "Ft ", RON: "lei ",
  ISK: "kr ", TRY: "\u20ba", JPY: "\u00a5", CNY: "\u00a5", HKD: "HK$", SGD: "S$",
  KRW: "\u20a9", INR: "\u20b9", IDR: "Rp", MYR: "RM", PHP: "\u20b1", THB: "\u0e3f",
  ILS: "\u20aa", AUD: "A$", NZD: "NZ$", CAD: "C$", BRL: "R$", MXN: "MX$",
  ZAR: "R",
};

export function symbolFor(currency: string) {
  return SYMBOLS[currency] ?? `${currency} `;
}

/** 2140.5 -> "2,140.50". Cents are dropped once a number gets long enough to crowd a phone. */
export function money(value: number, opts: { cents?: boolean } = {}) {
  const cents = opts.cents ?? Math.abs(value) < 100_000;
  return value.toLocaleString("en-US", {
    minimumFractionDigits: cents ? 2 : 0,
    maximumFractionDigits: cents ? 2 : 0,
  });
}

export function compact(value: number) {
  if (Math.abs(value) < 1000) return money(value, { cents: false });
  return value.toLocaleString("en-US", { notation: "compact", maximumFractionDigits: 1 });
}

/** Parses a date-only string without letting the local timezone shift the day. */
export function parseDay(iso: string) {
  const [y, m, d] = iso.split("-").map(Number);
  return new Date(y, (m ?? 1) - 1, d ?? 1);
}

export function todayISO() {
  const now = new Date();
  const pad = (n: number) => String(n).padStart(2, "0");
  return `${now.getFullYear()}-${pad(now.getMonth() + 1)}-${pad(now.getDate())}`;
}

const MONTHS = ["JAN","FEB","MAR","APR","MAY","JUN","JUL","AUG","SEP","OCT","NOV","DEC"];

export function shortDate(iso: string) {
  const d = parseDay(iso);
  return `${MONTHS[d.getMonth()]} ${d.getDate()}`;
}

export function dateRange(start: string | null, end: string | null) {
  if (!start && !end) return "Dates open";
  if (start && !end) return `From ${shortDate(start)}`;
  if (!start && end) return `Until ${shortDate(end!)}`;
  return `${shortDate(start!)} — ${shortDate(end!)}`;
}

export function dayLabel(iso: string) {
  const d = parseDay(iso);
  const today = parseDay(todayISO());
  const diff = Math.round((d.getTime() - today.getTime()) / 86_400_000);
  if (diff === 0) return "Today";
  if (diff === -1) return "Yesterday";
  return d.toLocaleDateString("en-US", { weekday: "short", month: "short", day: "numeric" });
}

export type TripPhase =
  | { kind: "upcoming"; days: number }
  | { kind: "active"; day: number; total: number }
  | { kind: "past" }
  | { kind: "undated" };

export function tripPhase(trip: { start_date: string | null; end_date: string | null }): TripPhase {
  const { start_date, end_date } = trip;
  if (!start_date && !end_date) return { kind: "undated" };
  const today = parseDay(todayISO()).getTime();
  const start = start_date ? parseDay(start_date).getTime() : today;
  const end = end_date ? parseDay(end_date).getTime() : start;

  if (today < start) return { kind: "upcoming", days: Math.ceil((start - today) / 86_400_000) };
  if (today > end) return { kind: "past" };
  return {
    kind: "active",
    day: Math.floor((today - start) / 86_400_000) + 1,
    total: Math.floor((end - start) / 86_400_000) + 1,
  };
}

/** Airport-style three letter code derived from the destination or trip name. */
export function tripCode(trip: { destination: string | null; name: string }) {
  const source = (trip.destination || trip.name).replace(/[^a-zA-Z ]/g, " ").trim();
  const words = source.split(/\s+/).filter(Boolean);
  if (words.length >= 3) return words.slice(0, 3).map((w) => w[0]).join("").toUpperCase();
  if (words.length === 2) return (words[0].slice(0, 2) + words[1][0]).toUpperCase();
  return (words[0]?.slice(0, 3) || "WYF").toUpperCase().padEnd(3, "X");
}

File: android/app/src/main/java/com/wayfare/app/core/Format.kt
package com.wayfare.app.core

import java.math.BigDecimal
import java.math.RoundingMode
import java.text.DecimalFormat
import java.text.DecimalFormatSymbols
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit
import java.util.Locale

/**
 * Money and date rendering, ported from `src/lib/format.ts`.
 *
 * The web app formats every figure with the en-US grouping style regardless of
 * the reader's locale, so the two clients have to agree here or the same trip
 * reads as €2,140.50 in the browser and € 2.140,50 on the phone.
 */

private val SYMBOLS = mapOf(
    "EUR" to "€", "USD" to "$", "GBP" to "£", "CHF" to "CHF ", "SEK" to "kr ",
    "NOK" to "kr ", "DKK" to "kr ", "PLN" to "zł ", "CZK" to "Kč ", "HUF" to "Ft ",
    "RON" to "lei ", "ISK" to "kr ", "TRY" to "₺", "JPY" to "¥", "CNY" to "¥",
    "HKD" to "HK$", "SGD" to "S$", "KRW" to "₩", "INR" to "₹", "IDR" to "Rp",
    "MYR" to "RM", "PHP" to "₱", "THB" to "฿", "ILS" to "₪", "AUD" to "A$",
    "NZD" to "NZ$", "CAD" to "C$", "BRL" to "R$", "MXN" to "MX$", "ZAR" to "R",
)

fun symbolFor(currency: String): String = SYMBOLS[currency] ?: "$currency "

private val US = DecimalFormatSymbols(Locale.US)
private val WITH_CENTS = DecimalFormat("#,##0.00", US)
private val WITHOUT_CENTS = DecimalFormat("#,##0", US)

/** 2140.5 -> "2,140.50". Cents are dropped once a number gets long enough to crowd a phone. */
fun amountText(value: BigDecimal, cents: Boolean = value.abs() < BigDecimal(100_000)): String =
    if (cents) WITH_CENTS.format(value) else WITHOUT_CENTS.format(value.setScale(0, RoundingMode.HALF_UP))

/** The full figure a reader sees, e.g. "€2,140.50" or "CHF 88.00". */
fun money(value: BigDecimal, currency: String = "EUR"): String =
    symbolFor(currency) + amountText(value)

// Sentence case, not capitals: the native design system reserves uppercase for
// the single 8sp superscript role, and these dates sit on every trip card.
private val MONTHS = listOf(
    "Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec",
)

fun shortDate(date: LocalDate): String = "${MONTHS[date.monthValue - 1]} ${date.dayOfMonth}"

fun dateRange(start: LocalDate?, end: LocalDate?): String = when {
    start == null && end == null -> "Dates open"
    end == null -> "From ${shortDate(start!!)}"
    start == null -> "Until ${shortDate(end)}"
    else -> "${shortDate(start)} — ${shortDate(end)}"
}

fun dayLabel(date: LocalDate, today: LocalDate = LocalDate.now()): String =
    when (ChronoUnit.DAYS.between(today, date)) {
        0L -> "Today"
        -1L -> "Yesterday"
        else -> date.format(DateTimeFormatter.ofPattern("EEE, MMM d", Locale.US))
    }
